### PR TITLE
Generics support

### PR DESCRIPTION
**Pull request by arendjr**
This PR implements basic support for generics, resolving #14.

I am still very new to implementing these kinds of macros, and I'm well aware it's not the prettiest yet, so advice is welcome!

### CODE REVIEW

File: example/src/lib.rs [hunk 1]
```diff
@@ -37,11 +37,28 @@ enum Vehicle {
     Car { brand: String, color: String },
 }
 
+#[derive(Serialize, TS)]
+struct Point<T>
+where
+    T: TS,
```
Comment (arendjr) [inline example/src/lib.rs:43]: I would have hoped we didn't need to specify this `where` clause manually, but I felt rewriting the `where` clause was a little bit too much to hope for in this first version...

File: example/src/lib.rs [hunk 2]
```diff
@@ -37,11 +37,28 @@ enum Vehicle {
...
+{
+    time: u64,
+    value: T,
+}
+
+#[derive(Serialize, TS)]
+struct Series {
+    points: Vec<Point<u64>>,
```
Comment (arendjr) [inline example/src/lib.rs:51]: This one deserves a test on its own. While `Point<u64>` now passes the tests, embedding it inside a vector actually causes subtle breakage (the generic argument is no longer passed to TS). This might be useful for a follow-up PR, as I felt getting all edge-cases right in one go is currently out of my league.

File: example/src/lib.rs [hunk 3]
```diff
@@ -37,11 +37,28 @@ enum Vehicle {
...
+}
+
 // this will export [Role] to `role.ts` and [User] to `user.ts` when running `cargo test`.
 // `export!` will also take care of including imports in typescript files.
 export! {
     Role => "role.ts",
     User => "user.ts",
+    // any type can be used here in place of the generic, as long as it impls TS:
```
Comment (arendjr) [inline example/src/lib.rs:59]: This feels a little bit like a work-around, but it's not our biggest problem :sweat_smile:

File: macros/src/types/named.rs [hunk 4]
```diff
@@ -102,3 +179,18 @@ fn format_field(
...
+        || ident == "BTreeMap"
+        || ident == "BTreeSet"
+        || ident == "Cell"
+        || ident == "Cow"
+        || ident == "HashMap"
+        || ident == "HashSet"
+        || ident == "Option"
+        || ident == "Rc"
+        || ident == "RefCell"
+        || ident == "Vec"
```
Comment (arendjr) [inline macros/src/types/named.rs:195]: This looks really ugly, but I'm at a loss for how we can detect this at compile time and still have a graceful fallback...

File: example/src/lib.rs (hunk 1 above)
Comment (arendjr) [reply example/src/lib.rs:43]: On second thought, if we did automatic rewriting, we wouldn't just need to rewrite it in the struct definition, but also for the `impl` blocks. That seems unreasonable, and I think it would also be very intuitive if you need to do manually for the `impl` blocks and *not* for the struct definition. So I'm starting to think it should stay like this.

**Comment by arendjr** [comment]
Superseded by #18.
